src/Tasks/Microsoft.NET.Build.Tasks/AppHost.cs

```diff
@@ -52,45 +52,54 @@ internal static class AppHost
+                    using (MemoryMappedViewAccessor accessor = memoryMappedFile.CreateViewAccessor())
+                    {
+                        SearchAndReplace(accessor, AppBinaryPathPlaceholderSearchValue, bytesToWrite, appHostSourceFilePath);
 
-                    appHostIsPEImage = IsPEImage(accessor);
+                        appHostIsPEImage = IsPEImage(accessor);
 
-                    if (windowsGraphicalUserInterface)
-                    {
-                        if (!appHostIsPEImage)
+                        if (windowsGraphicalUserInterface)
                         {
-                            throw new BuildErrorException(Strings.AppHostNotWindows, appHostSourceFilePath);
-                        }
+                            if (!appHostIsPEImage)
+                            {
+                                throw new BuildErrorException(Strings.AppHostNotWindows, appHostSourceFilePath);
+                            }
 
-                        SetWindowsGraphicalUserInterfaceBit(accessor, appHostSourceFilePath);
+                            SetWindowsGraphicalUserInterfaceBit(accessor, appHostSourceFilePath);
+                        }
                     }
                 }
-            }
 
-            if (intermediateAssembly != null && appHostIsPEImage)
-            {
-                if (ResourceUpdater.IsSupportedOS())
-                {
-                    // Copy resources from managed dll to the apphost
-                    new ResourceUpdater(appHostDestinationFilePath)
-                        .AddResourcesFromPEImage(intermediateAssembly)
-                        .Update();
-                }
-                else if (log != null)
+                if (intermediateAssembly != null && appHostIsPEImage)
                 {
-                    log.LogWarning(Strings.AppHostCustomizationRequiresWindowsHostWarning);
+                    if (ResourceUpdater.IsSupportedOS())
+                    {
+                        // Copy resources from managed dll to the apphost
+                        new ResourceUpdater(appHostDestinationFilePath)
+                            .AddResourcesFromPEImage(intermediateAssembly)
+                            .Update();
+                    }
+                    else if (log != null)
+                    {
+                        log.LogWarning(Strings.AppHostCustomizationRequiresWindowsHostWarning);
+                    }
                 }
-            }
 
-            // Memory-mapped write does not updating last write time
-            File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+                // Memory-mapped write does not updating last write time
+                File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+            }
+            catch (Exception)
+            {
+                // Delete the destination file so we don't leave an unmodified apphost
+                File.Delete(appHostDestinationFilePath);
```

Comment:
nit: what if this throws, say because the current user doesn't have permission to write or delete the file?

Reply:
I'll catch the specific exceptions and log them if a logger was provided.